File: occasum-core/src/main/java/pw/stamina/occasum/node/AbstractPropertyNode.java
package pw.stamina.occasum.node;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.factory.PropertyNodeFactory;
import pw.stamina.occasum.node.visit.PropertyNodeVisitor;
import pw.stamina.occasum.properties.Property;

import java.util.*;
import java.util.function.Predicate;

public abstract class AbstractPropertyNode implements PropertyNode {
    private final PropertyNodeFactory factory;

    protected final PropertyHandle handle;
    private List<PropertyNode> children;

    protected AbstractPropertyNode(PropertyNodeFactory factory,
                                   PropertyHandle handle) {
        Objects.requireNonNull(factory, "factory");
        Objects.requireNonNull(handle, "handle");

        this.factory = factory;
        this.handle = handle;
    }

    @Override
    public final PropertyHandle getHandle() {
        return handle;
    }

    @Override
    public final boolean hasChildren() {
        return children != null;
    }

    @Override
    public final List<PropertyNode> findChildren() {
        if (hasChildren()) {
            return Collections.unmodifiableList(children);
        }

        return Collections.emptyList();
    }

    @Override
    public final PropertyNode property(PropertyHandle handle, Property property) {
        String propertyId = property.getId();

        if (find(propertyId).isPresent()) {
            throw new IllegalArgumentException(
                    "another node with the same id has " +
                            "already been registered. Id: " + propertyId);
        }

        return add(factory.property(handle, this, property));
    }

    @Override
    public final PropertyNode property(Property property) {
        return property(handle, property);
    }

    @Override
    public final PropertyNode folder(PropertyHandle handle, String name) {
        return find(name).orElseGet(() -> add(factory.folder(handle, this, name)));
    }

    @Override
    public final PropertyNode folder(String name) {
        return folder(handle, name);
    }

    private PropertyNode add(PropertyNode node) {
        createChildrenIfAbsent();

        children.add(node);
        return node;
    }

    private void createChildrenIfAbsent() {
        if (!hasChildren()) {
            children = new ArrayList<>();
        }
    }

    @Override
    public void removeChild(PropertyNode node) {
        if (hasChildren() && children.remove(node)) {
            factory.notifyRemoval(node);
        }
    }

    @Override
    public void removeChild(String id) {
        find(id).ifPresent(this::removeChild);
    }

    @Override
    public final Optional<PropertyNode> find(String id) {
        if (!hasChildren()) {
            return Optional.empty();
        }

        return children.stream()
                .filter(doesIdMatchIgnoreCase(id))
                .findAny();
    }

    private static Predicate<PropertyNode> doesIdMatchIgnoreCase(String id) {
        return node -> node.getId().equalsIgnoreCase(id);
    }

    @Override
    public void accept(PropertyNodeVisitor visitor) {
        visitor.visitNode(this);

        if (hasChildren()) {
            visitChildren(visitor);
        }
    }

    private void visitChildren(PropertyNodeVisitor visitor) {
        children.forEach(childNode -> visitor.visitChildNode(childNode).ifPresent(childNode::accept));
    }

    @Override
    public String toString() {
        return "PropertyNode{id=" + getId() + '}';
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/AbstractPropertyNode.kt
package pw.stamina.occasum.node

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.node.factory.PropertyNodeFactory
import pw.stamina.occasum.node.visit.PropertyNodeVisitor
import pw.stamina.occasum.properties.Property

abstract class AbstractPropertyNode protected constructor(
        private val factory: PropertyNodeFactory,
        override val handle: PropertyHandle
) : PropertyNode {

    override val children: MutableList<PropertyNode> by lazy(LazyThreadSafetyMode.NONE) {
        ArrayList<PropertyNode>()
    }

    override fun property(handle: PropertyHandle, property: Property<*>): PropertyNode {
        val propertyId = property.id

        require(this[propertyId] == null) {
            "another node with the same id ('$propertyId') has already been registered"
        }

        return add(factory.property(handle, this, property))
    }

    override fun property(property: Property<*>): PropertyNode {
        return property(handle, property)
    }

    override fun folder(handle: PropertyHandle, name: String): PropertyNode {
        return this[name] ?: add(factory.folder(handle, this, name))
    }

    override fun folder(name: String): PropertyNode {
        return folder(handle, name)
    }

    private fun add(node: PropertyNode): PropertyNode {
        children.add(node)
        return node
    }

    override fun removeChild(node: PropertyNode) {
        if (children.remove(node)) {
            factory.notifyRemoval(node)
        }
    }

    override fun removeChild(id: String) {
        val child = this[id] ?: return
        removeChild(child)
    }

    override operator fun get(id: String): PropertyNode? {
        return children.find(doesIdMatchIgnoreCase(id))
    }

    private fun doesIdMatchIgnoreCase(id: String): (PropertyNode) -> Boolean {
        return { node -> node.id.equals(id, true) }
    }

    override fun accept(visitor: PropertyNodeVisitor) {
        visitor.visitNode(this)
        visitChildren(visitor)
    }

    private fun visitChildren(visitor: PropertyNodeVisitor) {
        children.forEach { childNode ->
            visitor.visitChildNode(childNode)?.let { childNode.accept(it) }
        }
    }

    override fun toString(): String {
        return "PropertyNode{id=$id}"
    }
}
